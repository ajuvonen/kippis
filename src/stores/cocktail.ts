import {
  getByIngredients,
  getBySearchString,
  getDetails,
  getNonAlcoholic,
  getByFirstLetter,
} from '@/api';
import {SEARCHABLE_ALCOHOLS} from '@/utils/constants';
import type {FullDetailsCocktail, SearchResultCocktail, SearchStringProps} from '@/utils/types';
import {defineStore} from 'pinia';
import {compose, flatten, map, prop, sortBy, uniq, without} from 'ramda';

export const useCocktailStore = defineStore('cocktail', {
  state: () => ({
    currentSearch: {} as SearchStringProps,
    cocktailDetails: [] as FullDetailsCocktail[],
    preventFetch: false,
    searchResults: [] as SearchResultCocktail[],
    selection: new Set<number>(),
    highlightedCocktail: null as number | null,
  }),
  getters: {
    getCocktailDetails: (state) => (id: number) => state.cocktailDetails.find((cocktail) => cocktail.id === id) || null,
    // Get a "shopping list" of ingredients. They should be unique, sorted and without some obvious items
    getAllIngredients(state): string[] {
      return compose(
        sortBy(prop(0)),
        without(['water', 'ice']),
        uniq,
        flatten<string[][]>,
        map((id: number) =>
          // Make cream, egg, and citrus fruits show up as their base form 
          this.getCocktailDetails(id)!.ingredients.map(({ingredient}) =>
            ingredient.replace(/^whipped | peel$| spiral$| white$| yolk$/, ''),
          ),
        ),
      )(Array.from(state.selection));
    },
  },
  actions: {
    async addToSelection(id: number) {
      await this.fetchCocktail(id);
      this.selection.add(id);
    },
    async fetchCocktail(id: number) {
      if (!this.cocktailDetails.some((cocktail) => cocktail.id === id)) {
        this.cocktailDetails.push(await getDetails(id));
      }
    },
    async openCocktailModal(id: number) {
      await this.fetchCocktail(id);
      this.highlightedCocktail = id;
    },
    removeFromSelection(id: number) {
      this.selection.delete(id);
    },
    async search(searchString: string) {
      if (searchString.length === 1) {
        this.searchResults = await getByFirstLetter(searchString);
      } else {
        this.searchResults = await getBySearchString(searchString);
      }
    },
    async searchNonAlcoholic() {
      this.searchResults = await getNonAlcoholic();
    },
    async searchWithTag(tag: string) {
      this.searchResults = await getByIngredients(
        SEARCHABLE_ALCOHOLS.find((alcohol) => tag === alcohol.tag)?.ingredients!,
      );
    },
  },
});
