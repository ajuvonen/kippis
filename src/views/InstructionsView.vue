<script setup lang="ts">
import {computed} from 'vue';
import {storeToRefs} from 'pinia';
import {intersection} from 'remeda';
import {useCocktailStore} from '@/stores/cocktail';
import {ALCOHOLS, FRUITS, MIXERS} from '@/utils/constants';
import {joinIngredients, randomDegree} from '@/utils/helpers';
import useScreen from '@/hooks/screenSize';
import SelectedCocktails from '@/components/SelectedCocktails.vue';
import SelectedCocktailsMobile from '@/components/SelectedCocktailsMobile.vue';
import LazyCocktailImage from '@/components/LazyCocktailImage.vue';
import CapitalizedList from '@/components/CapitalizedList.vue';
import SearchField from '@/components/SearchField.vue';

const cocktailStore = useCocktailStore();
const {getAllIngredients, selection} = storeToRefs(cocktailStore);

const {isSmallScreen} = useScreen();

const alcohols = computed(() => intersection(getAllIngredients.value, ALCOHOLS));
const mixers = computed(() => intersection(getAllIngredients.value, MIXERS));
const fruits = computed(() => intersection(getAllIngredients.value, FRUITS));
const other = computed(() =>
  getAllIngredients.value.filter(
    (ingredient) =>
      !alcohols.value.includes(ingredient) &&
      !mixers.value.includes(ingredient) &&
      !fruits.value.includes(ingredient),
  ),
);
</script>
<template>
  <SelectedCocktailsMobile v-if="isSmallScreen" />
  <SelectedCocktails v-else />
  <main class="flex-1">
    <h1>{{ $t('instructionsView.title') }}</h1>
    <div v-if="selection.length" class="flex">
      <div class="flex-1">
        <h2>{{ $t('instructionsView.ingredients') }}</h2>
        <CapitalizedList
          :title="$t('instructionsView.alcohols')"
          :items="alcohols"
          lang="en"
        ></CapitalizedList>
        <CapitalizedList
          :title="$t('instructionsView.mixers')"
          :items="mixers"
          lang="en"
        ></CapitalizedList>
        <CapitalizedList
          :title="$t('instructionsView.fruits')"
          :items="fruits"
          lang="en"
        ></CapitalizedList>
        <CapitalizedList
          :title="$t('instructionsView.other')"
          :items="other"
          lang="en"
        ></CapitalizedList>
        <h2>{{ $t('instructionsView.instructions') }}</h2>
        <div v-for="cocktail in selection" :key="cocktail.id" lang="en">
          <CapitalizedList
            :title="cocktail.name"
            :items="joinIngredients(cocktail.ingredients)"
          ></CapitalizedList>
          <p v-if="cocktail.instructions" class="mt-2">{{ cocktail.instructions }}</p>
        </div>
      </div>
      <div class="instructions__cocktail-image-container">
        <figure
          v-for="cocktail in selection"
          :key="cocktail.id"
          :style="`transform: rotate(${randomDegree()}deg);`"
          class="instructions__cocktail-image-wrapper"
        >
          <LazyCocktailImage :src="cocktail.thumb" class="instructions__cocktail-image" />
          <figcaption class="instructions__cocktail-image-label">
            {{ cocktail.name }}
          </figcaption>
        </figure>
      </div>
    </div>
    <div v-else>
      <p class="text-center my-4">{{ $t('instructionsView.noCocktails') }}</p>
      <SearchField :omit-title="true" />
    </div>
  </main>
</template>
<style scoped>
.instructions__cocktail-image-container {
  @apply hidden lg:flex flex-col w-1/3 ml-4;
}

.instructions__cocktail-image-wrapper {
  @apply shadow-sm;
}

.instructions__cocktail-image {
  @apply border-8 border-b-0 border-white;

  &:first-child {
    @apply mt-4;
  }
}

.instructions__cocktail-image-label {
  @apply bg-white h-12 p-2 text-center whitespace-nowrap overflow-hidden overflow-ellipsis text-3xl;
  font-family: 'Ephesis';
}

@media print {
  .instructions__cocktail-image-container {
    @apply block;
  }

  .instructions__cocktail-image {
    @apply break-inside-avoid;
  }
}
</style>
