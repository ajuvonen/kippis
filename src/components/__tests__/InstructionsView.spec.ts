import {describe, it, expect, beforeEach, vi, afterAll, beforeAll} from 'vitest';
import {mount} from '@vue/test-utils';
import {useCocktailStore} from '@/stores/cocktail';
import InstructionsView from '@/views/InstructionsView.vue';
import {testCocktails} from '@/components/__tests__/mswHandlers';

describe('InstructionsView', () => {
  let cocktailStore: ReturnType<typeof useCocktailStore>;

  beforeAll(() => {
    vi.mock('@/utils/helpers', async (importOriginal) => {
      const helpers = await importOriginal<typeof import('@/utils/helpers')>();
      return {
        ...helpers,
        randomDegree: () => 0,
      };
    });
  });

  beforeEach(() => {
    cocktailStore = useCocktailStore();
  });

  afterAll(() => {
    vi.restoreAllMocks();
  });

  it('mounts', () => {
    const wrapper = mount(InstructionsView);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('displays the side bar', () => {
    cocktailStore.selection.push(testCocktails[0][1]);
    const wrapper = mount(InstructionsView);
    expect(wrapper.find('.selected-cocktails').exists()).toBe(true);
  });

  it('displays the instructions', () => {
    cocktailStore.selection.push(testCocktails[0][1], testCocktails[1][1]);
    const wrapper = mount(InstructionsView);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
