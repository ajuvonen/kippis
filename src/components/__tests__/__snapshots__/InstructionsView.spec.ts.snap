// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InstructionsView > displays the instructions 1`] = `
"<transition-stub data-v-418939cb="" data-v-384a5d4b="" enterfromclass="ml-[-400px]" entertoclass="ml-0" leavefromclass="ml-0" leavetoclass="ml-[-400px]" enteractiveclass="transition-all" leaveactiveclass="transition-all" appear="false" persisted="false" css="true">
  <aside data-v-418939cb="" class="selected-cocktails">
    <h2 class="text-3xl text-center">Your Cocktails</h2>
    <ul data-v-a73318e6="" class="search-results pt-2 overflow-y-scroll">
      <li data-v-a73318e6="">
        <div data-v-cd13f5f7="" data-v-a73318e6="" class="search-result">
          <div data-v-cd13f5f7="" aria-label="View info about &quot;Margarita&quot;" class="search-result__image-wrapper" role="button" tabindex="0">
            <div data-v-35e45732="" data-v-cd13f5f7="" class="search-result__image">
              <div data-v-35e45732="" class="lazy-image__fallback"><svg data-v-35e45732="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
                  <path d="M7.5,7L5.5,5H18.5L16.5,7M11,13V19H6V21H18V19H13V13L21,5V3H3V5L11,13Z" class="transition-all"></path>
                </svg></div>
            </div>
            <div data-v-cd13f5f7="" class="search-result__shadow">Margarita</div>
          </div><button data-v-634d56d3="" data-v-cd13f5f7="" aria-label="Remove &quot;Margarita&quot; from selection" title="Remove &quot;Margarita&quot; from selection" class="remove-cocktail-0 action-button bg-rose-400 border-slate-800"><svg data-v-634d56d3="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
              <path d="M9,3V4H4V6H5V19A2,2 0 0,0 7,21H17A2,2 0 0,0 19,19V6H20V4H15V3H9M9,8H11V17H9V8M13,8H15V17H13V8Z" class="transition-all"></path>
            </svg></button>
        </div>
      </li>
      <li data-v-a73318e6="">
        <div data-v-cd13f5f7="" data-v-a73318e6="" class="search-result">
          <div data-v-cd13f5f7="" aria-label="View info about &quot;Mojito&quot;" class="search-result__image-wrapper" role="button" tabindex="0">
            <div data-v-35e45732="" data-v-cd13f5f7="" class="search-result__image">
              <div data-v-35e45732="" class="lazy-image__fallback"><svg data-v-35e45732="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
                  <path d="M7.5,7L5.5,5H18.5L16.5,7M11,13V19H6V21H18V19H13V13L21,5V3H3V5L11,13Z" class="transition-all"></path>
                </svg></div>
            </div>
            <div data-v-cd13f5f7="" class="search-result__shadow">Mojito</div>
          </div><button data-v-634d56d3="" data-v-cd13f5f7="" aria-label="Remove &quot;Mojito&quot; from selection" title="Remove &quot;Mojito&quot; from selection" class="remove-cocktail-1 action-button bg-rose-400 border-slate-800"><svg data-v-634d56d3="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
              <path d="M9,3V4H4V6H5V19A2,2 0 0,0 7,21H17A2,2 0 0,0 19,19V6H20V4H15V3H9M9,8H11V17H9V8M13,8H15V17H13V8Z" class="transition-all"></path>
            </svg></button>
        </div>
      </li>
    </ul><a href="/search" class="link-button flex items-center flex-shrink-0" data-test-id="selected-cocktails__action-button">Thirsty? Return to search.</a>
  </aside>
</transition-stub>
<main data-v-384a5d4b="" class="flex-1">
  <h1 data-v-384a5d4b="">Instructions and Ingredients</h1>
  <div data-v-384a5d4b="" class="flex">
    <div data-v-384a5d4b="" class="flex-1">
      <h2 data-v-384a5d4b="">Ingredients</h2>
      <div data-v-384a5d4b="" class="capitalized-list" lang="en">
        <h3>Alcohols</h3>
        <ul>
          <li class="text-sm uppercase leading-8">tequila</li>
          <li class="text-sm uppercase leading-8">triple sec</li>
        </ul>
      </div>
      <div data-v-384a5d4b="" class="capitalized-list" lang="en">
        <h3>Mixers</h3>
        <ul>
          <li class="text-sm uppercase leading-8">lime juice</li>
          <li class="text-sm uppercase leading-8">soda water</li>
        </ul>
      </div>
      <div data-v-384a5d4b="" class="capitalized-list" lang="en">
        <h3>Fruits and vegetables</h3>
        <ul>
          <li class="text-sm uppercase leading-8">lime</li>
          <li class="text-sm uppercase leading-8">mint</li>
        </ul>
      </div>
      <div data-v-384a5d4b="" class="capitalized-list" lang="en">
        <h3>Other ingredients</h3>
        <ul>
          <li class="text-sm uppercase leading-8">salt</li>
          <li class="text-sm uppercase leading-8">sugar</li>
        </ul>
      </div>
      <h2 data-v-384a5d4b="">Instructions</h2>
      <div data-v-384a5d4b="" lang="en">
        <div data-v-384a5d4b="" class="capitalized-list">
          <h3>Margarita</h3>
          <ul>
            <li class="text-sm uppercase leading-8">3 CL tequila</li>
            <li class="text-sm uppercase leading-8">1 CL triple sec</li>
            <li class="text-sm uppercase leading-8">3 CL lime juice</li>
            <li class="text-sm uppercase leading-8">salt</li>
          </ul>
        </div>
        <p data-v-384a5d4b="" class="mt-2">Rub the rim of the glass with the lime slice to make the salt stick to it.</p>
      </div>
      <div data-v-384a5d4b="" lang="en">
        <div data-v-384a5d4b="" class="capitalized-list">
          <h3>Mojito</h3>
          <ul>
            <li class="text-sm uppercase leading-8">Juice of 1 lime</li>
            <li class="text-sm uppercase leading-8">2 tsp sugar</li>
            <li class="text-sm uppercase leading-8">2-4 mint</li>
            <li class="text-sm uppercase leading-8">6 CL soda water</li>
          </ul>
        </div>
        <p data-v-384a5d4b="" class="mt-2">Muddle mint leaves with sugar and lime juice. Add a splash of soda water and fill the glass with cracked ice.</p>
      </div>
    </div>
    <div data-v-384a5d4b="" class="instructions__cocktail-image-container">
      <figure data-v-384a5d4b="" style="transform: rotate(0deg);" class="instructions__cocktail-image-wrapper">
        <div data-v-35e45732="" data-v-384a5d4b="" class="instructions__cocktail-image">
          <div data-v-35e45732="" class="lazy-image__fallback"><svg data-v-35e45732="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
              <path d="M7.5,7L5.5,5H18.5L16.5,7M11,13V19H6V21H18V19H13V13L21,5V3H3V5L11,13Z" class="transition-all"></path>
            </svg></div>
        </div>
        <figcaption data-v-384a5d4b="" class="instructions__cocktail-image-label">Margarita</figcaption>
      </figure>
      <figure data-v-384a5d4b="" style="transform: rotate(0deg);" class="instructions__cocktail-image-wrapper">
        <div data-v-35e45732="" data-v-384a5d4b="" class="instructions__cocktail-image">
          <div data-v-35e45732="" class="lazy-image__fallback"><svg data-v-35e45732="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
              <path d="M7.5,7L5.5,5H18.5L16.5,7M11,13V19H6V21H18V19H13V13L21,5V3H3V5L11,13Z" class="transition-all"></path>
            </svg></div>
        </div>
        <figcaption data-v-384a5d4b="" class="instructions__cocktail-image-label">Mojito</figcaption>
      </figure>
    </div>
  </div>
</main>"
`;

exports[`InstructionsView > mounts 1`] = `
"<transition-stub data-v-418939cb="" data-v-384a5d4b="" enterfromclass="ml-[-400px]" entertoclass="ml-0" leavefromclass="ml-0" leavetoclass="ml-[-400px]" enteractiveclass="transition-all" leaveactiveclass="transition-all" appear="false" persisted="false" css="true">
  <!--v-if-->
</transition-stub>
<main data-v-384a5d4b="" class="flex-1">
  <h1 data-v-384a5d4b="">Instructions and Ingredients</h1>
  <div data-v-384a5d4b="">
    <p data-v-384a5d4b="" class="text-center my-4">No cocktails selected. Use the search and find your favorites!</p>
    <div data-v-b87fbe66="" data-v-384a5d4b=""><label data-v-b87fbe66="" for="search-input" class="search-field__label sr-only">Search for Recipes</label>
      <div data-v-b87fbe66="" class="search-field__wrapper"><input data-v-b87fbe66="" id="search-input" type="search" maxlength="20" class="search-field__input"><button data-v-b87fbe66="" class="search-field__search-button"><svg data-v-b87fbe66="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
            <path d="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z" class="transition-all"></path>
          </svg><span data-v-b87fbe66="">Search</span></button></div>
      <ul data-v-b87fbe66="" class="search-field__tag-container" aria-label="Recipe Categories">
        <li data-v-b87fbe66=""><a data-v-b87fbe66="" href="/search?tag=gin" class="link-button flex items-center">Gin</a></li>
        <li data-v-b87fbe66=""><a data-v-b87fbe66="" href="/search?tag=rum" class="link-button flex items-center">Rum</a></li>
        <li data-v-b87fbe66=""><a data-v-b87fbe66="" href="/search?tag=vodka" class="link-button flex items-center">Vodka</a></li>
        <li data-v-b87fbe66=""><a data-v-b87fbe66="" href="/search?tag=whiskey" class="link-button flex items-center">Whiskey</a></li>
        <li data-v-b87fbe66=""><a data-v-b87fbe66="" href="/search?tag=virgin" class="link-button flex items-center">Virgin</a></li>
        <li data-v-b87fbe66=""><button data-v-b87fbe66="" data-test-id="search-field__random-button"><svg data-v-b87fbe66="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
              <path d="M14 8C13.45 8 13 7.55 13 7S13.45 6 14 6C14.55 6 15 6.45 15 7C15 7.55 14.55 8 14 8M12 12V19H5V12H12M12.78 10H4.22C3.55 10 3 10.55 3 11.22V19.78C3 20.45 3.55 21 4.22 21H12.78C13.45 21 14 20.45 14 19.78V11.22C14 10.55 13.45 10 12.78 10M19.78 3H11.22C10.55 3 10 3.55 10 4.22V8H12V5H19V12H16V14H19.78C20.45 14 21 13.45 21 12.78V4.22C21 3.55 20.45 3 19.78 3M17 8C16.45 8 16 7.55 16 7S16.45 6 17 6C17.55 6 18 6.45 18 7C18 7.55 17.55 8 17 8M17 11C16.45 11 16 10.55 16 10S16.45 9 17 9C17.55 9 18 9.45 18 10C18 10.55 17.55 11 17 11M7 15C6.45 15 6 14.55 6 14S6.45 13 7 13C7.55 13 8 13.45 8 14C8 14.55 7.55 15 7 15M10 18C9.45 18 9 17.55 9 17S9.45 16 10 16C10.55 16 11 16.45 11 17C11 17.55 10.55 18 10 18" class="transition-all"></path>
            </svg><span data-v-b87fbe66="">I'm Feeling Lucky</span></button></li>
      </ul>
    </div>
  </div>
</main>"
`;
